### ISSUE TITLE
Modify service setup URL method to rely on its v2 version

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#4376]
Related to #2044 and #2045, this is somewhat of a prerequisite for both of those. Please read the description and Acceptance Criteria for #2044 for additional context. This issue has been broken out to keep #2044 from getting too large of an issue, but it may not seem very useful on its own, which is why having an understanding of #2044 for this will help understanding the required changes here.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

Background: There is now a new version of the `/site-management/setup` service endpoint called `/v2/site-management/setup/`. Due to the changes further outlined in #2044, the new endpoint no longer has a welcome screen. This also means that the new endpoint is only ever relevant to be used for sites that are already registered on the service (i.e. site credentials have already been generated) and for users for who a temporary service session has been already initiated (this happens on the service under the hood when you get to the OAuth access screen).

The actual setup route will no longer be used directly when starting the setup flow. The only times when it'll be relevant going forward is when the user needs to be redirected back to the setup (e.g. during the redirect where data like the verification token is sent from the service to the plugin) or when an error occurred during the setup where the user needs to be sent back there.

There is a [POC](https://github.com/google/site-kit-wp/compare/poc/service-setup-v2-migration) which technically has all changes required here (and also most from #2044 and #2045) already implemented, however (since it's a POC) in partly quite hacky and unclean approaches. It's worth reviewing that POC for potential clarification and inspiration though - you can also try the branch locally with the staging version of the service to test out the full experience.

* If the `serviceSetupV2` flag is enabled, there should be a new `Google_Proxy::setup_url_v2()`, a new version of `Google_Proxy::setup_url()` that lives alongside it for the time being. It should return the new version of the setup URL, with adjusted query parameters based on the requirements for that new URL (outlined below).
    * There must always be a `code` query parameter (that's the parameter that contains the service session identifier). This was not a requirement for the original setup endpoint - potentially it's worth throwing an exception here if it is not set (either from the `$query_params` parameter of its filter).
    * There must always be a `site_id` **or** `site_code` parameter (that's the parameter that identifies the site). If a site hasn't yet received its site credentials from the service, it'll use a `site_code`, otherwise it'll use the `site_id`. These two parameters aren't new either, what's new though is that now one of them _has_ to be there.
    * There can optionally be a `step` query parameter, to point to a specific step from the setup.
    * All other query parameters currently added by the methods (e.g. `scope`, `nonce`, `supports`, `hl`, ...) are no longer relevant for the new endpoint, i.e. they should not be included in the returned URL. These parameters are going to be exclusively handled via the approach from #2045 (being passed to the `Google_Proxy::OAUTH2_SITE_URI` endpoint).
        * The only exception are the `verify` and `verification_method` query parameters which are used exclusively during verification. The `Google_Proxy::setup_url_v2` doesn't need to worry about those, this point here is just to clarify that those will still need to be added to the URL where relevant.
    * The new method should no longer include the `googlesitekit_proxy_setup_url_params` filter.
* The `OAuth_Client::get_proxy_setup_url()` method should remain untouched, and it will no longer be relevant for V2. Any code that currently relies on it should, if the `serviceSetupV2` feature flag is enabled, call `Google_Proxy::setup_url_v2()` directly, or if needed, a custom more specialized wrapper method (e.g. useful for automatically adding the right query parameters based on an error).
* All usage of the `googlesitekit_proxy_setup_url_params` filter (_except_ the one in the `Analytics` class which will be removed later via #2044) should be removed. Any remaining usage of it (after #4383 / #4451 already got rid of some) should be replaced with another approach where the relevant query parameters get passed to the method instead.
* Since the v2 version of the endpoint requires an "access code" (`code`) to be provided, it needs to be ensured that anywhere that one of the setup URLs methods can be called without that `code` (e.g. triggering the exception mentioned above), this no longer happens. This is at least relevant for the `Authentication::get_authentication_oauth_error_notice()` method.

#### `setup_url_v2()`

We're going to create a new function called `setup_url_v2`, which is responsible for validating and adding query params to the V2 setup URI.

* Create a new function named `setup_url_v2`, which accepts an array of query `$params`.
* Within `setup_url_v2`:
  * Conditionally throw an exception if the query param `code` is not set. Additionally, throw an exception if neither `site_id` or `site_code` are set.
  * Copy the return statement from `setup_url`, but change `self::SETUP_URI` to `self::SETUP_URI_V2`. Add the const `SETUP_URI_V2` at the top of the class, with the value `'/v2/site-management/setup/'`.
* The resulting function should look something like this:
```php
	public function setup_url_v2( array $params = array() ) {
		if ( empty( $params['code'] ) ) {
			throw new Exception( __( 'Missing code parameter for setup URL.', 'google-site-kit' ) );
		}
		if ( empty( $params['site_id'] ) && empty( $params['site_code'] ) ) {
			throw new Exception( __( 'Missing site_id or site_code parameter for setup URL.', 'google-site-kit' ) );
		}

		return add_query_arg( $params, $this->url( self::SETUP_URI_V2 ) );
	}
```

#### `add_setup_step_from_error_code()`

Create a helper function called `add_setup_step_from_error_code` which accepts query `$params` and an `$error_code`, and adds the `step` param based on the value of `$error_code`, returning `$params`.
```php
	public function add_setup_step_from_error_code( $params, $error_code ) {
		switch ( $error_code ) {
			case 'missing_verification':
				$params['step'] = 'verification';
				break;
			case 'missing_delegation_consent':
				$params['step'] = 'delegation_consent';
				break;
			case 'missing_search_console_property':
				$params['step'] = 'search_console_property';
				break;
		}
		return $params;
	}

```

### Within `includes/Core/Authentication/Clients/OAuth_Client.php`:

* Replace this call to `get_proxy_setup_url`:
https://github.com/google/site-kit-wp/blob/1bafd2e6cdc62eda8e9836b4989f81100e02ba13/includes/Core/Authentication/Clients/OAuth_Client.php#L388
* With a conditional call to `setup_url_v2` or `get_proxy_setup_url` depending on whether `serviceSetupV2` is enabled. The query params `code`, `site_id` and `step` should be passed in to `setup_url_v2`. For example:
```php
	if ( Feature_Flags::enabled( 'serviceSetupV2' ) ) {
		$params = array(
			'code'    => $e->getAccessCode(),
			'site_id' => $this->credentials->get()['oauth2_client_id']
		);
		$params = $this->google_proxy->add_setup_step_from_error_code( $params, $e->getMessage() );
		$url = $this->google_proxy->setup_url_v2( $params );
	} else {
		$url = $this->get_proxy_setup_url( $e->getAccessCode(), $e->getMessage() );
	}
	wp_safe_redirect( $url );
```

* In `get_proxy_setup_url`, at the top of the function, throw an exception if the function is called when the `serviceSetupV2` feature is enabled.

### Within `includes/Core/Authentication/Authentication.php`:

Here, we just need to replicate the PoC behaviour within the function `get_authentication_oauth_error_notice`. However we should isolate the feature flag conditional to make things more explicit.

Referring to this line from the PoC:

https://github.com/google/site-kit-wp/blob/0e6f380313516412df5435bb255cb6083551512f/includes/Core/Authentication/Authentication.php#L1079

Extract the condition, something like this:

```js
	$access_code = $this->user_options->get( OAuth_Client::OPTION_PROXY_ACCESS_CODE );

	if ( Feature_Flags::enabled( 'serviceSetupV2' ) ) {
		$is_using_proxy = $this->credentials->using_proxy() && $access_code;
	} else {
		$is_using_proxy = $this->credentials->using_proxy();
	}

	if ( $is_using_proxy ) {
		// NOTE: We'll also want to conditionally call `setup_url_v2` here, see below.
		$setup_url = $auth_client->get_proxy_setup_url( $access_code );
```

Also, note that the order of the first two conditionals within that `if/else` sequence should be switched in order to fall back to `$this->get_connect_url` in the case that there is no access code.

* Then, replace the call to `get_proxy_setup_url` with a conditionally invoked `setup_url_v2`/`get_proxy_setup_url`, e.g.
```php
	if ( Feature_Flags::enabled( 'serviceSetupV2' ) ) {
		$params = array(
			'code' => $access_code,
			'site_id' => $this->credentials->get()['oauth2_client_id']
		);
		$setup_url = $this->google_proxy->setup_url_v2( $params );
	} else {
		$setup_url = $auth_client->get_proxy_setup_url( $access_code );
	}
```

### Within `includes/Core/Modules/Module.php`:

* Replace the call to `get_proxy_setup_url` with a conditionally invoked `setup_url_v2`/`get_proxy_setup_url`, e.g.
```php
	if ( Feature_Flags::enabled( 'serviceSetupV2' ) ) {
		$params = array(
			'code' => $e->getAccessCode(),
			'site_id' => $auth_client->credentials()->get()['oauth2_client_id']
		);
		$params = $this->google_proxy->add_setup_step_from_error_code( $params, $code );
		$reconnect_url = $this->google_proxy->setup_url_v2( $params );
	} else {
		$reconnect_url = $auth_client->get_proxy_setup_url( $e->getAccessCode(), $code );
	}
```

### Within `includes/Core/Authentication/Setup.php`:

* In `redirect_to_proxy`, replace the call to `get_proxy_setup_url` with a conditionally invoked `setup_url_v2`/`get_proxy_setup_url`:
https://github.com/google/site-kit-wp/blob/1bafd2e6cdc62eda8e9836b4989f81100e02ba13/includes/Core/Authentication/Setup.php#L185
```php
               if ( Feature_Flags::enabled( 'serviceSetupV2' ) ) {
                       $params['code'] = $code;
                       $url = $this->authentication->get_google_proxy()->setup_url_v2( $params );
               } else {
                       $url = $this->authentication->get_oauth_client()->get_proxy_setup_url( $code );
                       $url = add_query_arg( $params, $url );
               }
```
### Within `includes/Modules/Site_Verification.php`:

* Remove this call to add the `googlesitekit_proxy_setup_url_params` filter. The `verify` and `verification_method` params will instead be provided in `Setup_V1` / `Setup_V2` without using a filter.
https://github.com/google/site-kit-wp/blob/ebac7c54faaee97e2626bc6421ce3c9c54657650/includes/Modules/Site_Verification.php#L407-L408

### Within `includes/Core/Authentication/Setup_V1.php`:

* At the point where `handle_verification` is called, create the query params `verify` and `verification_method`:
https://github.com/google/site-kit-wp/blob/1bafd2e6cdc62eda8e9836b4989f81100e02ba13/includes/Core/Authentication/Setup_V1.php#L91
```php
	$proxy_query_params = array(
		'verify'              => 'true',
		'verification_method' => $verification_method,
	);
```
* Ensure these query params are passed to `$this->redirect_to_proxy()` when it is invoked later in the same function.

### Within `includes/Core/Authentication/Setup_V2.php`:

* Where `$this->redirect_to_proxy()` is called without the `site_code` param specified, ensure the `site_id` param is passed in (derived from credentials).
* Where `handle_verification` is called, ensure the `verify` and `verification_method` params are passed to `$this->redirect_to_proxy()`.

### Test Coverage

* Add tests for the new methods in `includes/Core/Authentication/Google_Proxy.php`.
* Fix any failing tests.
* More tests may need updating when the feature flag is removed at a later date.

## QA Brief

This ticket is similar to #4383 in that the V2 flow can't be verified through regular QA. The same QA brief is applicable here:

* Test the current setup flow **without** the `serviceSetupV2` feature flag enabled
    * Make sure to test this on a fresh site (never used with the proxy before) to ensure the verification mechanic works
    * Make sure to also test this on a site which has already been verified

`QA:Eng` – secondary review of implementation to make sure ACs are met and no regressions are introduced that might not be caught in normal QA.

## Changelog entry

* N/A

**Comment by felixarntz** [comment]
@FlicHollis This is probably a good one to plan for Sprint 63 as well (also see [Asana comment](https://app.asana.com/0/0/1200049674554718/1201389444977856/f)).

**Comment by tofumatt** [comment]
I'm fine with the IB approach here over the POC—the approach here in the IB more cleanly separates the V1 and V2 logic. The IB is ✅  from me but as @felixarntz wrote the original POC I'd like him to review it too just to make sure it's right and the changes to the POC in this IB make sense for our PHP code 🙂

**Comment by felixarntz** [comment]
@techanvil The IB mostly looks good, +1 on the separation. I'd argue we can clean that up a bit further though since we are creating a new method infrastructure now. Some potential ideas/approaches:
* The new version of `Google_Proxy::setup_url` should never be called without an access code (`code` query parameter), so maybe we should clarify that through the method signature. It already requires the `Credentials` instance today (which is also used only to populate some query parameters), so IMO it would be good to require the access code via a distinct required method parameter too, and leave `$query_params` for the truly optional stuff.
* Alternatively to the above, one could argue that the new versions of `Google_Proxy::setup_url` and `OAuth_Client::get_proxy_setup_url` should have a clearer point of separation. Right now both of them contain logic, but I wonder if it would make more sense for all logic to be encapsulated in `OAuth_Client::get_proxy_setup_url` while `Google_Proxy::setup_url` is reduced to only receive query parameters and do nothing more than checking if required parameters are there (e.g. remove all the parts where it populates query parameters through other data).
* In either way, I wonder if we can make the flow of those methods easier to follow, specifically remove the `googlesitekit_proxy_setup_url_params` filter. It's confusing that query parameters can be altered via a call to the method but also the filter. The filter is only used in two places right now, which via #4383 will actually be differentiated more clearly, so both of those filter usages could probably be replaced by a direct call to get the proxy URL and redirect there, which would allow us to remove the filter entirely.

There is probably some legacy complexity involved in here, so I'm happy to jump on a call to discuss. Also looping in @aaemnnosttv who also has some context on this and maybe some additional ideas.

**Comment by techanvil** [comment]
Thanks @felixarntz, I'll take a look at the points you've raised, and might ask you to jump on a call to discuss, thanks :+1:

**Comment by felixarntz** [comment]
Thanks @techanvil @FlicHollis! Just confirming here that per our call this is now depending on #4383, so that one should be in the current sprint instead of this one.

**Comment by eugene-manuilov** [comment]
> The IB mostly looks good, +1 on the separation. I'd argue we can clean that up a bit further though since we are creating a new method infrastructure now.

@felixarntz do we really need to separate those methods? There are not that many changes in those methods to create separate versions of them. Plus, it's against our current approach in other places - we don't create a new react component with feature-specific changes, we make conditional changes in the component itself based on whether the feature flag is enabled or not. I think it would be better to keep one version of those methods with conditional changes as you have it in your POC - it is easier to read, understand and maintain.

> In setup_url_v2, after the lines above, conditionally throw an exception if code is not set. Additionally, throw an exception if neither site_id or site_code are set. For example:

@techanvil we should avoid throwing exceptions because it can lead to unexpected fatal errors with very bad user experience when something goes wrong. We should return an instance of the `WP_Error` class instead of throwing an exception and then die gracefully telling users what happens using the `wp_die` function.

**Comment by techanvil** [comment]
> @techanvil we should avoid throwing exceptions because it can lead to unexpected fatal errors with very bad user experience when something goes wrong. We should return an instance of the `WP_Error` class instead of throwing an exception and then die gracefully telling users what happens using the `wp_die` function.

@eugene-manuilov, thanks for pointing this out. When I read the point about throwing an exception in the AC, I searched the codebase for instances of exceptions being thrown and did find that happening in a number of places. Are these signs of an older approach, or do we sometimes throw exceptions but only when the `throw/catch` behaviour is desired?

**Comment by eugene-manuilov** [comment]
> @eugene-manuilov, thanks for pointing this out. When I read the point about throwing an exception in the AC, I searched the codebase for instances of exceptions being thrown and did find that happening in a number of places. Are these signs of an older approach, or do we sometimes throw exceptions but only when the `throw/catch` behaviour is desired?

@techanvil those had been added before I joined the project, so I can't say anything about them. But now going forward we should avoid doing it in my opinion.

**Comment by felixarntz** [comment]
@eugene-manuilov Regarding splitting the methods, that's a fair point, but I think it's a good approach here to simplify the code, and it sets us up for the future when the feature flag will be removed - which for this epic likely won't take a long time as it's fairly uncontroversial.

Good point about the exception. At a high level I agree that an exception is the worst experience for end users, but also exceptions are reasonable to throw in places where the code should never even get to in regular usage. Us developers who build Site Kit should see such errors with "highest" alert possible, there should be no chance to miss them - if a user runs into an exception like this, it should be considered a critical bug in our plugin. Some high-level differentiation between how I see it:
* **Throw `Exception` when:** Something is broken from incorrect _code_ usage, e.g. our own or a 3P developer's plugin code calling a method in our plugin the wrong way. This has to be fixed ASAP.
* **Use `wp_die()` when:** Something is broken from incorrect _user_ usage, e.g. a user manually navigating to a URL they shouldn't be able to get to, or a user manually removing some required query parameter from a URL. This does not have to be fixed.
* **Use `_doing_it_wrong()` when:** Something is "just" a problem that is worth fixing, but doesn't necessarily prevent a major functionality from working correctly and is not an access-related problem. This has to be fixed soon.

So from what I can tell here this issue would fall under the `Exception` group. Of course the above differentiation is subject to discussion, but that's how I see it.

**Comment by eugene-manuilov** [comment]
> In setup_url_v2, after the lines above, conditionally throw an exception if `code` is not set. Additionally, throw an exception if neither `site_id` or `site_code` are set.

@felixarntz, if `code`, `site_id` or `site_code` parameters are not set or empty - this is not incorrect _code_ usage. These values come from external sources (`code` comes as a GET param from the proxy server, `site_id` and `site_code` from the database) and in some circumstances there might be edge cases. For example, recently we had a ticket where a problem was caused by an incorrectly configured Nginx server that stripped query arguments from URLs.

**Comment by felixarntz** [comment]
@eugene-manuilov Yes, the `code` comes from a query parameter, but if that query parameter is not present, then the code that gets the query parameter should already `wp_die()` - it shouldn't just act as if that parameter was there and call the method, because if it did that _would_ be incorrect code usage. `site_code` is the same, that also comes from a query parameter. If we didn't want to manually `wp_die()` there and proceed without any conditional existence checks, then we could wrap the call to get the setup URL into a try-catch and `wp_die()` based on the exception. In that case, it's still incorrect code usage "invoked" by incorrect user usage.

Going forward (as of #4383 and #4376), all query parameters for `Google_Proxy::setup_url` will be passed directly into the method (no more filter), so the code that calls that method already should ensure that it doesn't call the method incorrectly.

**Comment by felixarntz** [comment]
@techanvil With #4451 now _almost_ through code review (only super-minor changes requested there), it may now be possible to proceed with an IB for this issue. It's probably best if you pull that PR's branch and come up with an IB based on that.

I've also made some small changes to the ACs to reflect the latest requirements, including removal of the `googlesitekit_proxy_setup_url_params` filter as we recently discussed.

**Comment by aaemnnosttv** [comment]
#4451 is now merged, so it's no longer a blocker here 🚀

**Comment by techanvil** [comment]
@felixarntz @aaemnnosttv that's great, thanks for letting me know. I'll move back onto this next!

**Comment by techanvil** [comment]
Hi @felixarntz, regarding the point about removing `googlesitekit_proxy_setup_url_params` - that looks like it might not be so practical now, due to the recent addition of this `googlesitekit_proxy_setup_url_params` filter in `includes/Modules/Analytics.php`:

https://github.com/google/site-kit-wp/blob/ebac7c54faaee97e2626bc6421ce3c9c54657650/includes/Modules/Analytics.php#L117

https://github.com/google/site-kit-wp/blob/ebac7c54faaee97e2626bc6421ce3c9c54657650/includes/Modules/Analytics.php#L1357-L1363

As introduced in this PR: https://github.com/google/site-kit-wp/pull/4411/files.

In order to replicate this behaviour without using the filter, we'd need to do something like this:

```php
		if ( $modules->is_module_active('analytics') && ! $modules->is_module_connected('analytics') ) {
			$query_params['mode'] = 'analytics-step';
		}
```

However there would be difficulty in accessing the `Modules` instance at the point where we construct the URL.

Say for example we are adding that `mode` query param within `Core/Authentication/Clients/OAuth_Client->get_proxy_setup_url()`. We'd need to pass `$modules` into the ` Core/Authentication/Authentication` instance, in order for the `Authentication` instance to be able to pass `$modules` into the `OAuth_Client` constructor. However, besides being somewhat unwieldy, this would be problematic as `Authentication` is actually a dependency of `Modules`, so passing the `Modules` instance to `Authentication` would introduce a circular dependency.

https://github.com/google/site-kit-wp/blob/ebac7c54faaee97e2626bc6421ce3c9c54657650/includes/Plugin.php#L161-L167

From my understanding of things, it looks like the `googlesitekit_proxy_setup_url_params` approach might be the best solution here. Again, please let me know if that needs any clarification or you'd like to jump on a call to discuss.

Interested to hear your thoughts! Cheers

**Comment by techanvil** [comment]
@felixarntz, further on the above, I have updated the IB, but have not attempted to remove the `googlesitekit_proxy_setup_url_params` filter. I've assigned this back over to you in order to get your thoughts on the filter question, prior to sending this back to IBR.

**Comment by felixarntz** [comment]
@techanvil Good point - I've updated the ACs to replace any `googlesitekit_proxy_setup_url_params` usage _other_ than that `Analytics` one, since that indeed cannot be removed until #2044, which is going to replace it. So the filter will then be removed via that issue.

Regarding the current IB:
* All proposed changes regarding the new `Google_Proxy::setup_url_v2` look good to me. Specifically, it doesn't contain the  `googlesitekit_proxy_setup_url_params` filter, which it indeed should not. The one usage in `Analytics` isn't relevant for the v2 setup URL anyway, since in V2 it will be provided through another way (that will be responsibility of #2044).
* Regarding `OAuth_Client::get_proxy_setup_url_v2`, I am now questioning whether we even need that method anymore for V2. It has always been confusing to have that method alongside the `Google_Proxy::setup_url` one which from its name looks like it would do exactly the same. With the `googlesitekit_proxy_setup_url_params` filter irrelevant for V2 (as mentioned above), there's only one thing this method would do, which is adding a `step` parameter based on an error code if one is provided. Also the overall method signature would be too limited as it wouldn't allow to pass any query parameters in like the underlying `Google_Proxy::setup_url_v2` allows for. I suggest the following instead:
    * Leave `OAuth_Client::get_proxy_setup_url` untouched as V1 legacy, and don't create a direct replacement for it.
    * In the few places where it is currently called, rely on `Google_Proxy::setup_url_v2` directly _if_ the `serviceSetupV2` feature flag is enabled.
        * If calling `Google_Proxy::setup_url_v2` directly in those places results in too much duplicate code for adding the right query parameters, let's introduce specialized helper methods with reusable logic. For example we'd probably need some sort of `get_setup_step_from_error_code( $error_code )` method for the part of determining the right `step` query parameter value based on a given error code.
    * Not strictly necessary, but since `OAuth_Client::get_proxy_setup_url` would now be legacy, it may be worth throwing an exception there if that method is ever called when the `serviceSetupV2` feature flag is active.
* The change in `Authentication` would need to be adjusted accordingly to the above on no longer having any `OAuth_Client::get_proxy_setup_url` method for V2.
* The `Site_Verification` change here is the only one that's not exactly correct: Those two parameters are actually still needed for V2, and I see now that the ACs may have been misleading about that. So at least for V2, we would need to go another route here, without that filter. Since we want to remove the filter in #2044 completely though, I'd say we should even for V1 replace the filter usage here with something else: Instead of the filter, the respective action handler method should conditionally add those same query parameters to the proxy URL to redirect to explicitly. Those actions are:
    * `Setup_V1::handle_action_setup`: The parameters should only be added to the URL if the `handle_verification` method was also called (since this action is a bit messy and has multiple purposes), so that it behaves the same as today.
    * `Setup_V2::handle_action_verify`: Here the parameters should be added unconditionally, since verification is the sole purpose of that method.
* Overall, the new `Setup` and `Setup_V2` classes introduced via #4383 probably also require a few more updates here, specifically to account for the above that we should not be using `OAuth_Client::get_proxy_setup_url` in V2.

I've updated the ACs a bit further to reflect all of the above. Apologies for the ongoing changes here - the existing code for V1 has been way too complicated, and having all the new enhancements scattered across 4 issues while they are all closely intertwined (specifically because of the V1 mess) has caused some mix-ups and oversights here and there. Nonetheless, we should do our best here to implement a clean V2 slate (which will then be especially clean once the V1 parts of the codebase will be removed, eventually) - I think not making the "double method" mistake of `OAuth_Client::get_proxy_setup_url` again is crucial here.

Let me know if that all makes sense to you - happy to jump on a call to clarify and discuss further.

**Comment by techanvil** [comment]
Thanks @felixarntz, that broadly makes sense, I'll see if I can figure out the IB, but may ask to jump on a call to flesh out some details depending on how it goes. Cheers

**Comment by techanvil** [comment]
Hi @felixarntz, regarding this point here:

> Regarding OAuth_Client::get_proxy_setup_url_v2, I am now questioning whether we even need that method anymore for V2. It has always been confusing to have that method alongside the Google_Proxy::setup_url one which from its name looks like it would do exactly the same. With the googlesitekit_proxy_setup_url_params filter irrelevant for V2 (as mentioned above), there's only one thing this method would do, which is adding a step parameter based on an error code if one is provided.

It's possible that you have overlooked the dependency on the `Credentials` instance for adding the `site_id` parameter - if we get rid of `get_proxy_setup_url_v2` entirely then we'd need to also pass in `site_id` or the `Credentials` instance everywhere we call `setup_url_v2`, which I think results in a bit more complexity than simply keeping `get_proxy_setup_url_v2` in some shape or form.

It might be useful to jump on a call to avoid too much back and forth here...I'll drop you a line on Slack to figure out the details.

**Comment by felixarntz** [comment]
@techanvil I'd prefer if we could figure out a way to add the `site_id` parameter in a more "manual" way. Of course without resulting in much duplicate code, but if we need to, we could have some specialized utility functions like I suggested for the `step` based on error code above.

Some context:
* In several circumstances where we need to call `setup_url_v2`, we could simply pass `site_id` as the ID from `Credentials` unconditionally, as there are a few places where the site already needs to be connected (i.e. `Credentials` is not "empty") to even reach that part of the code.
* The only place where `site_id` may not be present (yet) is in those action handlers in the new `Setup*` classes. We could probably have a simple protected method like `add_site_id_or_site_code` there that takes care of the logic:
    * If a site ID is available (via `Credentials`), set that as `site_id`.
    * Otherwise, set `site_code` using the site code from the incoming query parameter.
    * One of the two should always exist (e.g. the proxy passes the site code anytime the site doesn't have its site ID yet), otherwise something is broken :)

We can still set up a call on Monday if needed 👍

**Comment by eclarke1** [comment]
@felixarntz just an FYI that @techanvil is off sick today, however this ticket is now blocking the other 2 remaining tickets 2044 and 2045 - is there anything we can do to move this one along at all as I am conscious the other 2 tickets are quite large ones too. Thanks (cc @aaemnnosttv )

**Comment by techanvil** [comment]
Hi folks, just an update to say I'm back in and picking this up again. @felixarntz, thanks for the info - I'm hopeful this provides enough for me to finish the IB, but will let you know if it turns out a call would still be beneficial...

**Comment by techanvil** [comment]
Hi @felixarntz - I've updated the IB, I had momentarily unassigned the ticket from you but I guess it's probably best if I send it back over so you can take a look. I hope this is in line with your thinking. Cheers

**Comment by felixarntz** [comment]
@techanvil Thanks, the IB is almost good to go! A few minor last things:
> Apply the `googlesitekit_proxy_setup_url_params` filter.

That filter should no longer be applied in the `setup_url_v2` method, per the ACs - it should only remain in V1. The only remaining usage after your IB is implemented will be in the `Analytics` class, however that is even at this moment only relevant for V1, as with the new setup that parameter will be provided differently - all of that will be responsibility of #2044.

> Additionally, add an accessor `get_credentials`:

Hmm, isn't there already a method `Authentication::credentials()` that covers that use-case?

Can you check on the above? Once those have been addressed/clarified, I think this is good to go.

**Comment by techanvil** [comment]
Thanks @felixarntz - two good points there! I had a slight misunderstanding of the `googlesitekit_proxy_setup_url_params` and thought the `Analytics` usage was still valid in V2 for now, so thanks for clearing that up. Also, good spot on `credentials` - there is indeed an accessor of that name.

I've updated the IB to a) remove `googlesitekit_proxy_setup_url_params` from the `setup_url_v2`, and b) not instruct to add `get_credentials`, and replace usage of `get_credentials` with the existing `credentials` accessor.

So, hopefully, this will now be good to go! :crossed_fingers:

**Comment by felixarntz** [comment]
@techanvil Excellent, thanks! IB ✅

**Comment by wpdarren** [comment]
### QA Update: ✅

**Verified:**

- Tested the current setup flow without the `serviceSetupV2` feature flag enabled.
- Went through the set up on a fresh site and also new Google account.
- No issues occurred while setting up Site Kit or while connecting the modules.
- Went through the set up on a site which had been verified.
- No issues occurred while setting up Site Kit or while connecting the modules.

Added the `QA:Eng` label on for the second part of testing.

**Comment by felixarntz** [comment]
### QA ✅ 

Code changes follow the AC requirements, and functionality of V1 setup flow works the same as before.
